src/main/java/org/jenkinsci/plugins/docker/commons/DockerTool.java

```diff
@@ -66,16 +64,27 @@
      * Gets the executable name to use for a given launcher.
      * Suitable for the first item in {@link ArgumentListBuilder}.
      * @param name the name of the selected tool, or null for the default
-     * @param launcher a launcher specifying the slave (currently unused)
+     * @param node optionally, a node (such as a slave) on which we are running Docker
+     * @param listener a listener, required in case {@code node} is not null
+     * @param env optionally, environment variables to use when expanding the home directory
      * @return {@code docker} or an absolute path
      */
-    public static @Nonnull String getExecutable(@CheckForNull String name, @Nonnull Launcher launcher) {
+    public static @Nonnull String getExecutable(@CheckForNull String name, @CheckForNull Node node, @Nullable TaskListener listener, @CheckForNull EnvVars env) throws IOException, InterruptedException {
         if (name != null) {
             Jenkins j = Jenkins.getInstance();
             if (j != null) {
                 for (DockerTool tool : j.getDescriptorByType(DescriptorImpl.class).getInstallations()) {
                     if (tool.getName().equals(name)) {
-                        return tool.getExecutable(launcher);
+                        if (node != null) {
+                            tool = tool.forNode(node, listener);
+                        }
+                        if (env != null) {
+                            tool = tool.forEnvironment(env);
+                        }
```

Comment:
> there is some sort of overlaying going on here

Well, the tool can be node-specific (generally due to using a `ToolInstaller`, but also a `ToolLocationNodeProperty`). More rarely it can expand environment variables in `home`. They are orthogonal. This is just a standard pattern for any use of a `ToolInstallation`.